companies/views.py

```diff
@@ -1,3 +1,41 @@
-from django.shortcuts import render
+import json
 
-# Create your views here.
+from django.views     import View
+from django.http      import JsonResponse
+from django.db.models import Q
+
+from .models import NotificationTag, Image, Notification, Tag
+
+class NotificationlistView(View):
+    def get(self,request):
+        try:
+            tag = None
+            if request.GET.keys():
+                if 'tag' in request.GET.keys():
+                    tag    = request.GET['tag']
+                else:
+                    return JsonResponse({'MESSAGE':'KEY_ERROR'},status=400)
+
+            limit  = int(request.GET.get('limit',16))
+            offset = int(request.GET.get('offset',0))
+            q = Q()
+
+            if tag:
+                if not Tag.objects.filter(id=tag).exists():
+                    return JsonResponse({'MESSAGE': 'TAG_DOES_NOT_EXIST'},status=404)
+                q &= Q(notification_id=tag)
+
+            notification_tags = NotificationTag.objects.filter(q)
+            notification = [{
+                'title'   : notification_tag.notification.title,
+                'image'   : Image.objects.filter(notification_id=notification_tag.notification.id).first().image_url,
+                'id'      : notification_tag.notification.id,
+                'area'    : notification_tag.notification.company.address[:2],
+                'company' : notification_tag.notification.company.name
+                } for notification_tag in notification_tags]
+            return JsonResponse({'notification' : notification, 'total' : len(notification)}, status = 200)
```

Comment:
재연님!
Q객체 잘 사용해보셨군요! 👏 

그런데 tag 가 아닌 잘못된 query parameter 를 보냈을 경우 에러를 발생시키는 상황은 피해주세요.
 ```suggestion
        try:
            tag = None
            if request.GET.keys():
                if 'tag' in request.GET.keys():
                    tag    = request.GET['tag']
                else:
                    return JsonResponse({'MESSAGE':'KEY_ERROR'},status=400)    
```
예를 들어 프론트에서 'tag' 가 아닌 'category' 로 요청했다고 가정해봅시다.

현재와 같은 경우, 프론트는 서버로부터 에러에 대한 응답을 받게 됩니다. 여기서 재연님의 목적이 사용자에게 잘못된 요청이라는 것을 알려주는 것이 목적이라면, 프론트는 에러코드를 받은 이후 에러용 특정 페이지로 이동을 시키겠지요. 이 때 사용자는 불필요한 화면을 마주하게 됩니다. 
또 하나, 잘못된 요청(잘못된 query param) 을 보냈다 하더라도, product 의 리스트를 보여주는 것이 목적이라면, 프론트는 error 메시지를 받으면 다시 서버측에 product 리스트를 요청하겠지요. 즉 불필요한 요청을 한 번 더 발생시키게 됩니다.

이때는 error 발생 대신 None 을 받았을 때와 동일 하게 전체 리스트를 전달해주는 것이 적합합니다.

Reply:
확인햇습니다!
필터링을  하나만 구현하기 때문에 Q객체 -> filter 사용으로 바꿨습니다!!